src/algebra/commute.lean

```diff
@@ -0,0 +1,250 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Commuting pairs of elements in not-necessarily-commutative monoids,
+groups and (semi) rings.  Centralizers as submonoids, subgroups or
+subrings.
+
+Usage example: suppose that a, b and c are elements of a ring, and
+that `hb : commute a b` and `hc : commute a c`.  Then
+`hb.pow_left 5` proves `commute (a ^ 5) b` and
+`(hb.pow 2).add (hb.mul hc)` proves `commute a (b ^ 2 + b * c)`.
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(hb.pow_left 5).eq]`
+rather than just `rw [hb.pow_left 5]`.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+@[reducible] def commute (a b : S) : Prop := a * b = b * a
+
+theorem commute.eq {a b : S} (h : commute a b) : a * b = b * a := h
+
+@[refl] theorem commute.refl (a : S) : commute a a := by { unfold commute }
+
+@[symm] theorem commute.symm {a b : S} : commute a b → commute b a :=
+λ h, h.symm
+
+theorem commute.mul {a b c : S} :
+ commute a b → commute a c → commute a (b * c) :=
+λ hab hac,
+  by { dsimp [commute] at *,
+       rw [mul_assoc, ← hac, ← mul_assoc b, ← hab, mul_assoc] }
+
+theorem commute.mul_left {a b c : S} :
+ commute a c → commute b c → commute (a * b) c :=
+λ hac hbc, (hac.symm.mul hbc.symm).symm
+
+def centralizer (a : S) : set S := { x | commute a x }
+
+theorem mem_centralizer (a : S) {x : S} : x ∈ centralizer a ↔ commute a x :=
+iff.refl _
+
+def set_centralizer (T : set S) : set S :=
+  { x | ∀ (a : S), a ∈ T → commute a x }
+
+theorem mem_set_centralizer (T : set S) {x : S} :
+  x ∈ set_centralizer T ↔ ∀ (a : S), a ∈ T → commute a x :=
+iff.refl _
+
+end semigroup
+
+section monoid
+
+variables {M : Type*} [monoid M]
+
+theorem commute.one (a : M) : commute a 1 :=
+by { dsimp [commute], rw [one_mul, mul_one] }
+
+theorem commute.one_left (a : M) : commute 1 a := (commute.one a).symm
+
+theorem commute.pow {a b : M} (hab : commute a b) : ∀ (n : ℕ), commute a (b ^ n)
```

Comment:
If you prove `centralizer.is_submonoid` first, most lemmas about `pow` follow from the fact that submonoids are closed w.r.t `pow`.